crates/aptos-crypto/src/p256/p256_sigs.rs

```diff
@@ -0,0 +1,191 @@
+        if bytes.len() != P256_SIGNATURE_LENGTH {
+            return Err(CryptoMaterialError::WrongLengthError);
+        }
+        if !P256Signature::check_s_lt_order_half(&bytes[32..]) {
+            return Err(CryptoMaterialError::CanonicalRepresentationError);
+        }
+        Ok(())
+    }
+
+    /// Check if S < ORDER_HALF to capture invalid signatures.
+    fn check_s_lt_order_half(s: &[u8]) -> bool {
+        for i in 0..32 {
+            match s[i].cmp(&ORDER_HALF[i]) {
+                Ordering::Less => return true,
+                Ordering::Greater => return false,
+                _ => {},
+            }
+        }
+        // As this stage S == ORDER_HALF which implies a non canonical S.
+        false
+    }
+
+    /// If the signature {R,S} does not have S < n/2 where n is the Ristretto255 order, return
+    /// {R,n-S} as the canonical encoding of this signature to prevent malleability attacks. See
+    /// `check_s_malleability` for more detail 
+    pub fn make_canonical(&self) -> P256Signature {
+        if P256Signature::check_s_malleability(&self.to_bytes()[..]).is_ok() {
+            return self.clone()
+        };
+        let s = self.0.s();
+        let r = self.0.r();
+        // NonZeroScalar::try_from throws an error on being passed the curve order, so we use the
+        // order minus one, then add one later to compute (n-1) - s + 1 = n-s
+        let order_minus_one = NonZeroScalar::try_from(&ORDER_MINUS_ONE[..]).unwrap();
+        let one = NonZeroScalar::from_uint(<NistP256 as Curve>::Uint::ONE).unwrap();
+        // Dereferencing a NonZeroScalar makes it a Scalar, which implements subtraction
+        let new_s = *order_minus_one - *s + *one;
+        let new_s_nonzero = NonZeroScalar::new(new_s).unwrap();
+        let new_sig = p256::ecdsa::Signature::from_scalars(&r, &new_s_nonzero).unwrap();
+        P256Signature(new_sig)
+    }
+}
+
+//////////////////////
+// Signature Traits //
+//////////////////////
+
+impl Signature for P256Signature {
+    type SigningKeyMaterial = P256PrivateKey;
+    type VerifyingKeyMaterial = P256PublicKey;
+
+    /// Verifies that the provided signature is valid for the provided message, going beyond the
+    /// [NIST SP 800-186](https://csrc.nist.gov/publications/detail/sp/800-186/final) specification, to prevent scalar malleability as done in [BIP146](https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki).
+    fn verify<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+        public_key: &P256PublicKey,
+    ) -> Result<()> {
+        Self::verify_arbitrary_msg(self, &signing_message(message)?, public_key)
+    }
+
+    /// Checks that `self` is valid for an arbitrary &[u8] `message` using `public_key`.
+    /// Outside of this crate, this particular function should only be used for native signature
+    /// verification in Move.
+    ///
+    /// This function will check both the signature and `public_key` for small subgroup attacks.
```

Comment:
`secp256r1` is [a prime-order group](https://safecurves.cr.yp.to/www.secg.org/sec2-v2.pdf). There are no small subgroup attacks.

Can you change this to say "This function will check if the signature has been mauled."?

Reply:
Changed to "Checks for and rejects non-canonical signatures (r,s) where s > (n/2), where n is the group order"